EncoderDecoder

cmanzo opened the pull request:
This PR introduces EncoderDecoder networks
**Example:**
```
autoencoder = dl.EncoderDecoder(
    in_channels=1,
    encoder_channels=[16, 32, 64],
    decoder_channels=[32, 16],
    out_channels=1,
    out_activation=Sigmoid,
)
```
Bottleneck and Output blocks can be added, e.g., by:
```
autenc = autoencoder.create()
autenc.bottleneck_blocks.configure(
    dl.ConvolutionalNeuralNetwork,
    in_channels=64,
    hidden_channels=[128, 128],
    out_channels=64,
)
```
and offer flexibility to introduce VAE and Wasserstein.

cmanzo committed:
encdec

cmanzo committed:
test

cmanzo committed:
test

BenjaminMidtvedt commented on .github/workflows/encdec.py:
```diff
@@ -0,0 +1,289 @@
+    @property
+    def bottleneck_layer(self) -> LayerList[Layer]:
+        """Return the layers of the network. Equivalent to `.blocks.layer`."""
+        return self.bottleneck_blocks
```
should be `self.bottleneck_blocks.layer`?

cmanzo replied:
yes, fixed by defining bottleneck_blocks as a PLAN block

BenjaminMidtvedt commented on .github/workflows/encdec.py:
```diff
@@ -0,0 +1,289 @@
+    @property
+    def output_layer(self) -> LayerList[Layer]:
+        """Return the layers of the network. Equivalent to `.blocks.layer`."""
+        return self.output_blocks
```
check that this is correct too

cmanzo replied:
same as above

BenjaminMidtvedt commented on .github/workflows/encdec.py:
I think this file can be removed?

cmanzo replied:
removed

BenjaminMidtvedt commented:
In general I really like it. There are a few design decision that I think we should all join and decide on.

First `unpool` as the name for the upsampling layer. I quite like it, but we should all agree.

Second, for the decoder, we should probably have `UnpoolLayerActivationNormalization` block.

I'm also thinking that we want the order to be `LayerActivationNormalizationUnpool` instead. The reason being how skip connections are going to work. 

```py
def forward(self, x)
    down_1 = self.pool_layer_activation_block_1(x)
    down_2 = self.pool_layer_activation_block_2(down_1)

    up_1 = self.layer_activation_unpool(down_2)
    up_1 = nn.cat((down_1, up_1), dim=1)

    up_2 = self.layer_activation_unpool(up_1)
    up_2 = nn.cat((x, up_2), dim=1)

    return self.layer_activation(up_2)
```

Maybe we call it `ConvolutionalEncoderDecoder`? Maybe even `ConvolutionalEncoderDecoder2d`.

cmanzo committed:
encdec

cmanzo committed:
test

cmanzo committed:
remove workflow

cmanzo commented:
Regarding `unpool`, I also like it, @giovannivolpe what about it?

I have now changed the basic Encoder-Decoder architecture, using a `LayerActivationNormalizationUnpool` block for the decoder. The scheme is very symmetric, an autoencoder is defined as:

```import deeplay as dl
from torch.nn import Sigmoid

autoencoder = dl.ConvolutionalEncoderDecoder2d(
    in_channels=1,
    encoder_channels=[4, 8, 16],
    decoder_channels=[16, 8, 4],
    out_channels=1,
    out_activation=Sigmoid,
)
print(autoencoder)
```

Without specifying a bottleneck, we get pool-layer-activation-(bottleneck=identity)-layer-activation-unpool in the middle. The default output block only has a convolution and a custom activation.

I also changed the name to `ConvolutionalEncoderDecoder2d`.

giovannivolpe commented:
@cmanzo @BenjaminMidtvedt I'd go with "Upsample" because it's consistent with the book. Also Unpool assumes that a Pool has happened, which is not necessarily the case

cmanzo committed:
changed unpool to upsample - modules lowercase

cmanzo committed:
u

cmanzo committed:
trick to change case

cmanzo committed:
modules lower case in file name
